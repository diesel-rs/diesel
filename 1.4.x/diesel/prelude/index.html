<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Re-exports important traits and types. Meant to be glob imported when using Diesel."><meta name="keywords" content="rust, rustlang, rust-lang, prelude"><title>diesel::prelude - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg"><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../diesel/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><h2 class="location">Module prelude</h2><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li></ul></div><div id="sidebar-vars" data-name="prelude" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input"name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">diesel</a>::<wbr><a class="mod" href="#">prelude</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/diesel/lib.rs.html#336-364" title="goto source code">[src]</a></span></h1><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Re-exports important traits and types. Meant to be glob imported when using Diesel.</p>
</div></details><h2 id="reexports" class="section-header"><a href="#reexports">Re-exports</a></h2>
<div class="item-table"><div class="item-left import-item"><code>pub use associations::<a class="trait" href="../associations/trait.Identifiable.html" title="trait diesel::associations::Identifiable">Identifiable</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use connection::<a class="trait" href="../connection/trait.Connection.html" title="trait diesel::connection::Connection">Connection</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use deserialize::<a class="trait" href="../deserialize/trait.Queryable.html" title="trait diesel::deserialize::Queryable">Queryable</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use expression::<a class="trait" href="../expression/trait.AppearsOnTable.html" title="trait diesel::expression::AppearsOnTable">AppearsOnTable</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use expression::<a class="trait" href="../expression/trait.BoxableExpression.html" title="trait diesel::expression::BoxableExpression">BoxableExpression</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use expression::<a class="trait" href="../expression/trait.Expression.html" title="trait diesel::expression::Expression">Expression</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use expression::<a class="trait" href="../expression/trait.IntoSql.html" title="trait diesel::expression::IntoSql">IntoSql</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use expression::<a class="trait" href="../expression/trait.SelectableExpression.html" title="trait diesel::expression::SelectableExpression">SelectableExpression</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use <a class="mod" href="../expression_methods/index.html" title="mod diesel::expression_methods">expression_methods</a>::*;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use query_dsl::<a class="trait" href="../query_dsl/trait.QueryDsl.html" title="trait diesel::query_dsl::QueryDsl">QueryDsl</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use query_dsl::<a class="trait" href="../query_dsl/trait.RunQueryDsl.html" title="trait diesel::query_dsl::RunQueryDsl">RunQueryDsl</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use query_source::<a class="trait" href="../query_source/trait.Column.html" title="trait diesel::query_source::Column">Column</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use query_source::<a class="trait" href="../query_source/trait.QuerySource.html" title="trait diesel::query_source::QuerySource">QuerySource</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use query_source::<a class="trait" href="../query_source/trait.Table.html" title="trait diesel::query_source::Table">Table</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use result::<a class="enum" href="../result/enum.ConnectionError.html" title="enum diesel::result::ConnectionError">ConnectionError</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use result::<a class="type" href="../result/type.ConnectionResult.html" title="type diesel::result::ConnectionResult">ConnectionResult</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use result::<a class="trait" href="../result/trait.OptionalExtension.html" title="trait diesel::result::OptionalExtension">OptionalExtension</a>;</code></div><div class="item-right docblock-short"></div><div class="item-left import-item"><code>pub use result::<a class="type" href="../result/type.QueryResult.html" title="type diesel::result::QueryResult">QueryResult</a>;</code></div><div class="item-right docblock-short"></div></div><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="struct" href="struct.MysqlConnection.html" title="diesel::prelude::MysqlConnection struct">MysqlConnection</a></div><div class="item-right docblock-short"><p>A connection to a MySQL database. Connection URLs should be in the form
<code>mysql://[user[:password]@]host/database_name</code></p>
</div><div class="item-left module-item"><a class="struct" href="struct.PgConnection.html" title="diesel::prelude::PgConnection struct">PgConnection</a></div><div class="item-right docblock-short"><p>The connection string expected by <code>PgConnection::establish</code>
should be a PostgreSQL connection string, as documented at
<a href="https://www.postgresql.org/docs/9.4/static/libpq-connect.html#LIBPQ-CONNSTRING">https://www.postgresql.org/docs/9.4/static/libpq-connect.html#LIBPQ-CONNSTRING</a></p>
</div><div class="item-left module-item"><a class="struct" href="struct.SqliteConnection.html" title="diesel::prelude::SqliteConnection struct">SqliteConnection</a></div><div class="item-right docblock-short"><p>Connections for the SQLite backend. Unlike other backends, “connection URLs”
for SQLite are file paths, <a href="https://sqlite.org/uri.html">URIs</a>, or special
identifiers like <code>:memory:</code>.</p>
</div></div><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="trait" href="trait.BelongingToDsl.html" title="diesel::prelude::BelongingToDsl trait">BelongingToDsl</a></div><div class="item-right docblock-short"><p>Constructs a query that finds record(s) based on directional association with other record(s).</p>
</div><div class="item-left module-item"><a class="trait" href="trait.GroupedBy.html" title="diesel::prelude::GroupedBy trait">GroupedBy</a></div><div class="item-right docblock-short"><p>The <code>grouped_by</code> function groups records by their parent.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.Insertable.html" title="diesel::prelude::Insertable trait">Insertable</a></div><div class="item-right docblock-short"><p>Represents that a structure can be used to insert a new row into the
database. This is automatically implemented for <code>&amp;[T]</code> and <code>&amp;Vec&lt;T&gt;</code> for
inserting more than one record.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.JoinOnDsl.html" title="diesel::prelude::JoinOnDsl trait">JoinOnDsl</a></div><div class="item-right docblock-short"><p>Specify the <code>ON</code> clause for a join statement. This will override
any implicit <code>ON</code> clause that would come from <a href="../macro.joinable.html"><code>joinable!</code></a></p>
</div><div class="item-left module-item"><a class="trait" href="trait.JoinTo.html" title="diesel::prelude::JoinTo trait">JoinTo</a></div><div class="item-right docblock-short"><p>Indicates that two tables can be joined without an explicit <code>ON</code> clause.</p>
</div><div class="item-left module-item"><a class="trait" href="trait.SaveChangesDsl.html" title="diesel::prelude::SaveChangesDsl trait">SaveChangesDsl</a></div><div class="item-right docblock-short"><p>Sugar for types which implement both <code>AsChangeset</code> and <code>Identifiable</code></p>
</div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="diesel" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div>
    <script src="../../main.js"></script>
</body></html>