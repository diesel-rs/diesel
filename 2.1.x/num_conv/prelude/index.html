<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Anonymously import all extension traits."><title>num_conv::prelude - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="num_conv" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (e3181b091 2024-04-18)" data-channel="nightly" data-search-js="search-5b46ae87cc198d72.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-e32f0c247825364d.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../num_conv/index.html">num_conv</a><span class="version">0.1.0</span></h2></div><h2 class="location"><a href="#">Module prelude</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#reexports">Re-exports</a></li></ul></section><h2><a href="../index.html">In crate num_conv</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../num_conv/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">num_conv</a>::<wbr><a class="mod" href="#">prelude</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../src/num_conv/lib.rs.html#15">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Anonymously import all extension traits.</p>
<p>This allows you to use the methods without worrying about polluting the namespace or importing
them individually.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>num_conv::prelude::<span class="kw-2">*</span>;</code></pre></div>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport._"><code>pub use crate::<a class="trait" href="../trait.CastSigned.html" title="trait num_conv::CastSigned">CastSigned</a> as _;</code></div></li><li><div class="item-name" id="reexport._-1"><code>pub use crate::<a class="trait" href="../trait.CastUnsigned.html" title="trait num_conv::CastUnsigned">CastUnsigned</a> as _;</code></div></li><li><div class="item-name" id="reexport._-2"><code>pub use crate::<a class="trait" href="../trait.Extend.html" title="trait num_conv::Extend">Extend</a> as _;</code></div></li><li><div class="item-name" id="reexport._-3"><code>pub use crate::<a class="trait" href="../trait.Truncate.html" title="trait num_conv::Truncate">Truncate</a> as _;</code></div></li></ul></section></div></main></body></html>