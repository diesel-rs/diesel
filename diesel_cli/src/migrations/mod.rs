use chrono::Utc;
use clap::ArgMatches;
use diesel::backend::Backend;
use diesel::migration::{Migration, MigrationSource};
use diesel::Connection;
use diesel_migrations::{FileBasedMigrations, HarnessWithOutput, MigrationError, MigrationHarness};
use std::any::Any;
use std::collections::{HashMap, HashSet};
use std::env;
use std::error::Error;
use std::fmt::Display;
use std::fs::{self};
use std::path::{Path, PathBuf};

use crate::database::InferConnection;
use crate::{config::Config, handle_error, regenerate_schema_if_file_specified};

mod diff_schema;

pub(super) fn run_migration_command(
    matches: &ArgMatches,
) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
    match matches.subcommand().unwrap() {
        ("run", _) => {
            let mut conn = InferConnection::from_matches(matches);
            let dir = migrations_dir(matches).unwrap_or_else(handle_error);
            let dir = FileBasedMigrations::from_path(dir).unwrap_or_else(handle_error);
            run_migrations_with_output(&mut conn, dir)?;
            regenerate_schema_if_file_specified(matches)?;
        }
        ("revert", args) => {
            let mut conn = InferConnection::from_matches(matches);
            let dir = migrations_dir(matches).unwrap_or_else(handle_error);
            let dir = FileBasedMigrations::from_path(dir).unwrap_or_else(handle_error);
            if args.get_flag("REVERT_ALL") {
                revert_all_migrations_with_output(&mut conn, dir)?;
            } else {
                let number = args.get_one::<u64>("REVERT_NUMBER").unwrap();
                for _ in 0..*number {
                    match revert_migration_with_output(&mut conn, dir.clone()) {
                        Ok(_) => {}
                        Err(e) if e.is::<MigrationError>() => {
                            match e.downcast_ref::<MigrationError>() {
                                // If n is larger then the actual number of migrations,
                                // just stop reverting them
                                Some(MigrationError::NoMigrationRun) => break,
                                _ => return Err(e),
                            }
                        }
                        Err(e) => return Err(e),
                    }
                }
            }

            regenerate_schema_if_file_specified(matches)?;
        }
        ("redo", args) => {
            let mut conn = InferConnection::from_matches(matches);
            let dir = migrations_dir(matches).unwrap_or_else(handle_error);
            let dir = FileBasedMigrations::from_path(dir).unwrap_or_else(handle_error);
            redo_migrations(&mut conn, dir, args);
            regenerate_schema_if_file_specified(matches)?;
        }
        ("list", _) => {
            let mut conn = InferConnection::from_matches(matches);
            let dir = migrations_dir(matches).unwrap_or_else(handle_error);
            let dir = FileBasedMigrations::from_path(dir).unwrap_or_else(handle_error);
            list_migrations(&mut conn, dir)?;
        }
        ("pending", _) => {
            let mut conn = InferConnection::from_matches(matches);
            let dir = migrations_dir(matches).unwrap_or_else(handle_error);
            let dir = FileBasedMigrations::from_path(dir).unwrap_or_else(handle_error);
            let result = MigrationHarness::has_pending_migration(&mut conn, dir)?;
            println!("{result:?}");
        }
        ("generate", args) => {
            let migration_name = args.get_one::<String>("MIGRATION_NAME").unwrap();

            let (up_sql, down_sql) = if let Some(diff_schema) = args.get_one::<String>("SCHEMA_RS")
            {
                let config = Config::read(matches)?;
                let diff_schema = if diff_schema == "NOT_SET" {
                    config.print_schema.file.clone()
                } else {
                    Some(PathBuf::from(diff_schema))
                };
                if let Some(diff_schema) = diff_schema {
                    self::diff_schema::generate_sql_based_on_diff_schema(
                        config,
                        matches,
                        &diff_schema,
                    )?
                } else {
                    (String::new(), String::new())
                }
            } else {
                (String::new(), String::new())
            };
            let version = migration_version(args);
            let versioned_name = format!("{version}_{migration_name}");
            let migration_dir = migrations_dir(matches)
                .unwrap_or_else(handle_error)
                .join(versioned_name);
            fs::create_dir(&migration_dir).unwrap();

            match args
                .get_one::<String>("MIGRATION_FORMAT")
                .map(|s| s as &str)
            {
                Some("sql") => generate_sql_migration(
                    &migration_dir,
                    !args.get_flag("MIGRATION_NO_DOWN_FILE"),
                    up_sql,
                    down_sql,
                ),
                Some(x) => return Err(format!("Unrecognized migration format `{x}`").into()),
                None => unreachable!("MIGRATION_FORMAT has a default value"),
            }
        }
        _ => unreachable!("The cli parser should prevent reaching here"),
    };

    Ok(())
}

fn generate_sql_migration(path: &Path, with_down: bool, up_sql: String, down_sql: String) {
    use std::io::Write;

    let migration_dir_relative =
        crate::convert_absolute_path_to_relative(path, &env::current_dir().unwrap());

    let up_path = path.join("up.sql");
    println!(
        "Creating {}",
        migration_dir_relative.join("up.sql").display()
    );
    let mut up = fs::File::create(up_path).unwrap();
    up.write_all(b"-- Your SQL goes here\n").unwrap();
    up.write_all(up_sql.as_bytes()).unwrap();

    if with_down {
        let down_path = path.join("down.sql");
        println!(
            "Creating {}",
            migration_dir_relative.join("down.sql").display()
        );
        let mut down = fs::File::create(down_path).unwrap();
        down.write_all(b"-- This file should undo anything in `up.sql`\n")
            .unwrap();
        down.write_all(down_sql.as_bytes()).unwrap();
    }
}

fn migration_version<'a>(matches: &'a ArgMatches) -> Box<dyn Display + 'a> {
    matches
        .get_one::<String>("MIGRATION_VERSION")
        .map(|s| Box::new(s) as Box<dyn Display>)
        .unwrap_or_else(|| Box::new(Utc::now().format(crate::TIMESTAMP_FORMAT)))
}

fn migrations_dir_from_cli(matches: &ArgMatches) -> Option<PathBuf> {
    matches.get_one("MIGRATION_DIRECTORY").cloned().or_else(|| {
        matches
            .subcommand()
            .and_then(|s| migrations_dir_from_cli(s.1))
    })
}

pub fn run_migrations_with_output<Conn, DB>(
    conn: &mut Conn,
    migrations: FileBasedMigrations,
) -> Result<(), Box<dyn Error + Send + Sync + 'static>>
where
    Conn: MigrationHarness<DB> + Connection<Backend = DB> + 'static,
    DB: Backend,
{
    HarnessWithOutput::write_to_stdout(conn)
        .run_pending_migrations(migrations)
        .map(|_| ())
}

fn revert_all_migrations_with_output<Conn, DB>(
    conn: &mut Conn,
    migrations: FileBasedMigrations,
) -> Result<(), Box<dyn Error + Send + Sync + 'static>>
where
    Conn: MigrationHarness<DB> + Connection<Backend = DB> + 'static,
    DB: Backend,
{
    HarnessWithOutput::write_to_stdout(conn)
        .revert_all_migrations(migrations)
        .map(|_| ())
}

fn revert_migration_with_output<Conn, DB>(
    conn: &mut Conn,
    migrations: FileBasedMigrations,
) -> Result<(), Box<dyn Error + Send + Sync + 'static>>
where
    Conn: MigrationHarness<DB> + Connection<Backend = DB> + 'static,
    DB: Backend,
{
    HarnessWithOutput::write_to_stdout(conn)
        .revert_last_migration(migrations)
        .map(|_| ())
}

fn list_migrations<Conn, DB>(
    conn: &mut Conn,
    migrations: FileBasedMigrations,
) -> Result<(), Box<dyn Error + Send + Sync + 'static>>
where
    Conn: MigrationHarness<DB> + Connection<Backend = DB> + 'static,
    DB: Backend,
{
    let applied_migrations = conn
        .applied_migrations()?
        .into_iter()
        .collect::<HashSet<_>>();

    let mut migrations = MigrationSource::<DB>::migrations(&migrations)?;
    migrations.sort_unstable_by(|a, b| a.name().version().cmp(&b.name().version()));
    println!("Migrations:");
    for migration in migrations {
        let applied = applied_migrations.contains(&migration.name().version());
        let name = migration.name();
        let x = if applied { 'X' } else { ' ' };
        println!("  [{x}] {name}");
    }

    Ok(())
}

/// Checks for a migrations folder in the following order :
/// 1. From the CLI arguments
/// 2. From the MIGRATION_DIRECTORY environment variable
/// 3. From `diesel.toml` in the `migrations_directory` section
///
/// Else try to find the migrations directory with the
/// `find_migrations_directory` in the diesel_migrations crate.
///
/// Returns a `MigrationError::MigrationDirectoryNotFound` if
/// no path to the migration directory is found.
pub fn migrations_dir(matches: &ArgMatches) -> Result<PathBuf, MigrationError> {
    let migrations_dir = migrations_dir_from_cli(matches)
        .or_else(|| env::var("MIGRATION_DIRECTORY").map(PathBuf::from).ok())
        .or_else(|| {
            Some(
                Config::read(matches)
                    .unwrap_or_else(handle_error)
                    .migrations_directory?
                    .dir,
            )
        });

    match migrations_dir {
        Some(dir) => Ok(dir),
        None => FileBasedMigrations::find_migrations_directory().map(|p| p.path().to_path_buf()),
    }
}

/// Reverts all the migrations, and then runs them again, if the `--all`
/// argument is used. Otherwise it only redoes a specific number of migrations
/// if the `--number` argument is used.
/// We try to execute the migrations in a single transaction so that f either part fails,
/// the transaction is not committed.
/// If the list of migrations that need to be redone contains a single migration
/// with `run_in_transaction = false` or if the backend is MySQL we cannot use a
/// transaction.
fn redo_migrations<Conn, DB>(
    conn: &mut Conn,
    migrations_dir: FileBasedMigrations,
    args: &ArgMatches,
) where
    DB: Backend,
    Conn: MigrationHarness<DB> + Connection<Backend = DB> + 'static,
{
    let migrations = MigrationSource::<DB>::migrations(&migrations_dir)
        .unwrap_or_else(handle_error)
        .into_iter()
        .map(|m| (m.name().version().as_owned(), m))
        .collect::<HashMap<_, _>>();
    let applied_migrations = conn.applied_migrations().unwrap_or_else(handle_error);
    let versions_to_revert = if args.get_flag("REDO_ALL") {
        &applied_migrations
    } else {
        let number = args.get_one::<u64>("REDO_NUMBER").unwrap();
        let number = std::cmp::min(*number as usize, applied_migrations.len());
        &applied_migrations[..number]
    };
    let should_use_not_use_transaction = versions_to_revert.iter().any(|v| {
        migrations
            .get(v)
            .map(|m| !m.metadata().run_in_transaction())
            .unwrap_or_default()
    });

    let migrations_inner =
        |harness: &mut HarnessWithOutput<Conn, _>|
         -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
            // revert all the migrations
            let reverted_versions = if args.get_flag("REDO_ALL") {
                harness.revert_all_migrations(migrations_dir.clone())?
            } else {
                let number = args.get_one::<u64>("REDO_NUMBER").unwrap();
                (0..*number)
                    .filter_map(|_| {
                        match harness.revert_last_migration(migrations_dir.clone()) {
                            Ok(v) => Some(Ok(v)),
                            Err(e) if e.is::<MigrationError>() => {
                                match e.downcast_ref::<MigrationError>() {
                                    // If n is larger then the actual number of migrations,
                                    // just stop reverting them
                                    Some(MigrationError::NoMigrationRun) => None,
                                    _ => Some(Err(e)),
                                }
                            }
                            Err(e) => Some(Err(e)),
                        }
                    })
                    .collect::<Result<Vec<_>, _>>()?
            };

            // get a mapping between migrations and migration versions
             let mut migrations = MigrationSource::<DB>::migrations(&migrations_dir)
                 .unwrap_or_else(handle_error)
                 .into_iter()
                 .map(|m| (m.name().version().as_owned(), m))
                 .collect::<HashMap<_, _>>();

            // build a list of migrations that need to be applied
            let mut migrations = reverted_versions
                .into_iter()
                .map(|v| {
                    migrations
                        .remove(&v)
                        .ok_or_else(|| MigrationError::UnknownMigrationVersion(v.as_owned()))
                })
                .collect::<Result<Vec<_>, _>>()?;

            // Sort the migrations by version to apply them in order.
            migrations.sort_by_key(|m| m.name().version().as_owned());

            // apply all outstanding migrations
            harness.run_migrations(&migrations)?;

            Ok(())
        };

    if !should_use_not_use_transaction && should_redo_migration_in_transaction(conn) {
        conn.transaction(|conn| migrations_inner(&mut HarnessWithOutput::write_to_stdout(conn)))
            .unwrap_or_else(handle_error);
    } else {
        migrations_inner(&mut HarnessWithOutput::write_to_stdout(conn))
            .unwrap_or_else(handle_error);
    }
}

#[cfg(feature = "mysql")]
fn should_redo_migration_in_transaction(t: &dyn Any) -> bool {
    !matches!(
        t.downcast_ref::<InferConnection>(),
        Some(InferConnection::Mysql(_))
    )
}

#[cfg(not(feature = "mysql"))]
fn should_redo_migration_in_transaction(_t: &dyn Any) -> bool {
    true
}
