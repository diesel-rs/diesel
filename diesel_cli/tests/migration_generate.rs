use regex::Regex;
use chrono::prelude::*;

use support::project;
use diesel::migrations::TIMESTAMP_FORMAT;

#[test]
fn migration_generate_creates_a_migration_with_the_proper_name() {
    let p = project("migration_name")
        .folder("migrations")
        .build();
    let result = p.command("migration")
        .arg("generate")
        .arg("hello")
        .run();

    // check overall output
    let expected_stdout = Regex::new("\
Creating migrations.\\d*_hello.up.sql
Creating migrations.\\d*_hello.down.sql\
        ").unwrap();
    assert!(result.is_success(), "Command failed: {:?}", result);
    assert!(expected_stdout.is_match(result.stdout()));

    // check timestamps are properly formatted
    let captured_timestamps = Regex::new(r"(?P<stamp>\d*)_hello").unwrap();
    let mut stamps_found = 0;
    for caps in captured_timestamps.captures_iter(result.stdout()) {
        let timestamp = Utc.datetime_from_str(&caps["stamp"], TIMESTAMP_FORMAT);
        assert!(timestamp.is_ok(), "Found invalid timestamp format: {:?}", &caps["stamp"]);
        stamps_found += 1;
    }
    assert_eq!(stamps_found, 2);

    let migrations = p.migrations();
    assert_eq!(1, migrations.len());

    let migration = &migrations[0];
    assert_eq!("hello", migration.name());
    assert!(migration.path().join("up.sql").exists());
    assert!(migration.path().join("down.sql").exists());
}

#[test]
fn migration_generate_creates_a_migration_with_initital_contents() {
    let p = project("migration_name")
        .folder("migrations")
        .build();
    let result = p.command("migration")
        .arg("generate")
        .arg("hello")
        .run();
    assert!(result.is_success(), "Command failed: {:?}", result);

    let migrations = p.migrations();
    let migration = &migrations[0];

    let up = file_content(&migration.path().join("up.sql"));
    let down = file_content(&migration.path().join("down.sql"));

    assert_eq!(up.trim(), "-- Your SQL goes here");
    assert_eq!(down.trim(), "-- This file should undo anything in `up.sql`");

    fn file_content<P: AsRef<::std::path::Path>>(path: P) -> String {
        use std::io::Read;

        let mut file = ::std::fs::File::open(path).unwrap();
        let mut contents = String::new();
        file.read_to_string(&mut contents).unwrap();

        contents
    }
}

#[test]
fn migration_generate_doesnt_require_database_url_to_be_set() {
    let p = project("migration_name")
        .folder("migrations")
        .build();
    let result = p.command_without_database_url("migration")
        .arg("generate")
        .arg("hello")
        .run();

    assert!(result.is_success(), "Command failed: {:?}", result);
}

#[test]
fn migration_version_can_be_specified_on_creation() {
    let p = project("migration_name")
        .folder("migrations")
        .build();
    let result = p.command("migration")
        .arg("generate")
        .arg("hello")
        .arg("--version=1234")
        .run();

    let expected_stdout = Regex::new("\
Creating migrations.1234_hello.up.sql
Creating migrations.1234_hello.down.sql
").unwrap();
    assert!(result.is_success(), "Command failed: {:?}", result);
    assert!(expected_stdout.is_match(result.stdout()));

    assert!(p.has_file("migrations/1234_hello/up.sql"));
    assert!(p.has_file("migrations/1234_hello/down.sql"));
}

#[test]
fn migration_directory_can_be_specified_for_generate_by_command_line_arg() {
    let p = project("migration_name")
        .folder("foo")
        .build();
    let result = p.command("migration")
        .arg("generate")
        .arg("stuff")
        .arg("--version=12345")
        .arg("--migration-dir=foo")
        .run();

    let expected_stdout = Regex::new("\
Creating foo.12345_stuff.up.sql
Creating foo.12345_stuff.down.sql
").unwrap();
    assert!(result.is_success(), "Command failed: {:?}", result);
    assert!(expected_stdout.is_match(result.stdout()));

    assert!(p.has_file("foo/12345_stuff/up.sql"));
    assert!(p.has_file("foo/12345_stuff/down.sql"));
}

#[test]
fn migration_directory_can_be_specified_for_generate_by_env_var() {
    let p = project("migration_name")
        .folder("bar")
        .build();
    let result = p.command("migration")
        .arg("generate")
        .arg("stuff")
        .arg("--version=12345")
        .env("MIGRATION_DIRECTORY", "bar")
        .run();

    let expected_stdout = Regex::new("\
Creating bar.12345_stuff.up.sql
Creating bar.12345_stuff.down.sql
").unwrap();
    assert!(result.is_success(), "Command failed: {:?}", result);
    assert!(expected_stdout.is_match(result.stdout()));

    assert!(p.has_file("bar/12345_stuff/up.sql"));
    assert!(p.has_file("bar/12345_stuff/down.sql"));
}
