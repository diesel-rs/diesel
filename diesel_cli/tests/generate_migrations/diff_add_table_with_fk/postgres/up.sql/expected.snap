---
source: diesel_cli/tests/migration_generate.rs
assertion_line: 326
description: "Test: diff_add_table_with_fk"
---
-- Your SQL goes here

CREATE TABLE "posts"(
	"post_id" INTEGER NOT NULL PRIMARY KEY,
	"title" TEXT NOT NULL,
	"body" TEXT,
	"foreign_key_user_id" INTEGER NOT NULL,
	FOREIGN KEY ("foreign_key_user_id") REFERENCES "users"("user_id")
);


